### ISSUE TITLE
getByteData() and getThumbByteData(width: int, height: int, quality: int) both crashing the app

### ISSUE THREAD

**Issue by Yunus0or1** [Sh1d0w/multi_image_picker#358]
**Bug Report**

multi_image_picker: ^4.6.0-rc.3

Is it happening on Android: Yes

Is it happening on iOS: Do not know

```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, v1.12.13+hotfix.5, on Microsoft Windows [Version 10.0.18362.535], locale en-IN)

[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[√] Android Studio (version 3.5)
[√] VS Code (version 1.40.2)
[√] Connected device (1 available)

• No issues found!

```

**Describe the bug**

Accessing the image file via getByteData() crashes the app. Code is given below:

```
List<Asset> resultList;
  try {
        resultList = await MultiImagePicker.pickImages(
            maxImages: 10, enableCamera: true);
      } on Exception catch (e) {
        error = e as String;
      }
      for (int i = 0; i < resultList .length; i++) {
        try {
          ByteData imageData = await resultList[i].getThumbByteData(600, 800);
          // crashes here.
        } catch (error) {
          print(error);
        }
      }

```

But if I use `print(await images[i].isPortrait);` it produces output. 

Development Device: Samsung S5. 
minSdkVersion 19
targetSdkVersion 28
classpath 'com.android.tools.build:gradle:3.4.2'
classpath 'com.google.gms:google-services:4.2.0'

**Comment by Sh1d0w** [comment]
Please provide the crash stack trace and the error that you get, otherwise I can't help you figure out what the problem is.

**Comment by Yunus0or1** [comment]
Sorry to say, there is not even any crash stack tree. Just the app crashes and says Lost connection to the device.

**Comment by Sh1d0w** [comment]
To see the error you are getting, you need to run `adb logcat` in separate terminal then run your app. When it crashes look in the output of the logcat command and find the exact error you re getting and post it here.

**Comment by Yunus0or1** [comment]
As per I know flutter shows error in the console inside Run tab in Android Studio. And in Run tab it only shows "Lost connection to device." Nothing more. Usually I always debug issues from this Run tab. Any suggestion sir? 
![1](https://user-images.githubusercontent.com/11329052/71319325-09d0eb80-24c7-11ea-9b2e-d50902434ac6.PNG)

**Comment by Sh1d0w** [comment]
Yes, as I said you need to use [logcat](https://developer.android.com/studio/command-line/logcat) as the app crashes before you can see the logs in your run tab.

**Comment by Yunus0or1** [comment]
```

12-22 14:41:49.483  4619  4619 D StatusBar.NetworkController: refreshNwBoosterIndicator - setNWBoosterIndicators(false)
12-22 14:41:49.488  4619  4619 D StatusBar.NetworkController: refreshNwBoosterIndicator - setNWBoosterIndicators(false)
12-22 14:41:49.488  4619  4619 D StatusBar.NetworkController: refreshNwBoosterIndicator - setNWBoosterIndicators(false)
12-22 14:41:49.493  4619  4619 D StatusBar.NetworkController: refreshNwBoosterIndicator - setNWBoosterIndicators(false)
12-22 14:41:50.003 23407 23866 I art     : Starting a blocking GC Alloc
12-22 14:41:50.003 23407 23866 I art     : Starting a blocking GC Alloc
12-22 14:41:50.008 23407 23866 I art     : Alloc sticky concurrent mark sweep GC freed 792(34KB) AllocSpace objects, 0(0B) LOS objects, 15% free, 86MB/102MB, paused 261us total 5.438ms
12-22 14:41:50.008 23407 23866 I art     : Starting a blocking GC Alloc
12-22 14:41:50.028 23407 23866 I art     : Alloc partial concurrent mark sweep GC freed 2511(152KB) AllocSpace objects, 3(396KB) LOS objects, 15% free, 85MB/101MB, paused 491us total 17.641ms
12-22 14:41:50.028 23407 23866 I art     : Starting a blocking GC Alloc
12-22 14:41:50.053 23407 23866 I art     : Alloc concurrent mark sweep GC freed 229(23KB) AllocSpace objects, 0(0B) LOS objects, 15% free, 85MB/101MB, paused 526us total 26.195ms
12-22 14:41:50.053 23407 23866 I art     : Forcing collection of SoftReferences for 60MB allocation
12-22 14:41:50.053 23407 23866 I art     : Starting a blocking GC Alloc
12-22 14:41:50.073 23407 23866 I art     : Alloc concurrent mark sweep GC freed 3629(239KB) AllocSpace objects, 3(72KB) LOS objects, 15% free, 85MB/101MB, paused 383us total 21.179ms
12-22 14:41:50.073 23407 23866 W art     : Throwing OutOfMemoryError "Failed to allocate a 63489036 byte allocation with 16777216 free bytes and 42MB until OOM"
12-22 14:41:50.078 23407 23866 I art     : Starting a blocking GC Alloc
12-22 14:41:50.078 23407 23866 I art     : Starting a blocking GC Alloc
12-22 14:41:50.088 23407 23866 I art     : Alloc sticky concurrent mark sweep GC freed 3(384B) AllocSpace objects, 0(0B) LOS objects, 15% free, 85MB/101MB, paused 289us total 7.251ms
12-22 14:41:50.088 23407 23866 I art     : Starting a blocking GC Alloc
12-22 14:41:50.108 23407 23866 I art     : Alloc partial concurrent mark sweep GC freed 21(880B) AllocSpace objects, 0(0B) LOS objects, 15% free, 85MB/101MB, paused 378us total 19.472ms
12-22 14:41:50.108 23407 23866 I art     : Starting a blocking GC Alloc
12-22 14:41:50.128 23407 23866 I art     : Alloc concurrent mark sweep GC freed 10(416B) AllocSpace objects, 0(0B) LOS objects, 15% free, 85MB/101MB, paused 372us total 21.069ms
12-22 14:41:50.128 23407 23866 I art     : Forcing collection of SoftReferences for 60MB allocation
12-22 14:41:50.128 23407 23866 I art     : Starting a blocking GC Alloc
12-22 14:41:50.148 23407 23866 I art     : Alloc concurrent mark sweep GC freed 9(384B) AllocSpace objects, 0(0B) LOS objects, 15% free, 85MB/101MB, paused 305us total 19.074ms
12-22 14:41:50.148 23407 23866 W art     : Throwing OutOfMemoryError "Failed to allocate a 63489036 byte allocation with 16777216 free bytes and 42MB until OOM"
12-22 14:41:50.163 23407 23866 E AndroidRuntime: FATAL EXCEPTION: AsyncTask #4
12-22 14:41:50.163 23407 23866 E AndroidRuntime: Process: com.hurryup.merchant, PID: 23407
12-22 14:41:50.163 23407 23866 E AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at android.os.AsyncTask$3.done(AsyncTask.java:309)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at java.util.concurrent.FutureTask.run(FutureTask.java:242)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at java.lang.Thread.run(Thread.java:818)
12-22 14:41:50.163 23407 23866 E AndroidRuntime: Caused by: java.lang.OutOfMemoryError: Failed to allocate a 63489036 byte allocation with 16777216 free bytes and 42MB until OOM
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at dalvik.system.VMRuntime.newNonMovableArray(Native Method)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at android.graphics.Bitmap.nativeCreate(Native Method)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at android.graphics.Bitmap.createBitmap(Bitmap.java:975)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at android.graphics.Bitmap.createBitmap(Bitmap.java:946)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at android.graphics.Bitmap.createBitmap(Bitmap.java:877)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at com.vitanov.multiimagepicker.MultiImagePickerPlugin.getCorrectlyOrientedImage(MultiImagePickerPlugin.java:819)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at com.vitanov.multiimagepicker.MultiImagePickerPlugin.access$000(MultiImagePickerPlugin.java:70)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at com.vitanov.multiimagepicker.MultiImagePickerPlugin$GetThumbnailTask.doInBackground(MultiImagePickerPlugin.java:171)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at com.vitanov.multiimagepicker.MultiImagePickerPlugin$GetThumbnailTask.doInBackground(MultiImagePickerPlugin.java:143)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at android.os.AsyncTask$2.call(AsyncTask.java:295)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
12-22 14:41:50.163 23407 23866 E AndroidRuntime:        ... 4 more
12-22 14:41:50.163  4141  5417 W ActivityManager:   Force finishing activity com.hurryup.merchant/.MainActivity
12-22 14:41:50.163  4141  5417 D FocusedStackFrame: Set to : 0
12-22 14:41:50.173  4141  5417 D InputDispatcher: Focused application set to: xxxx
12-22 14:41:50.173  4141  5417 D InputDispatcher: Focus left window: 23407
12-22 14:41:50.178  4141  4331 D PointerIcon: setMouseIconStyle1 pointerType: 1001 iconType:101 flag:0 pid:4141 uid:1000
12-22 14:41:50.178  4141  4331 D PointerIcon: setMouseCustomIcon IconType is same.101
12-22 14:41:50.178  4141  4331 D PointerIcon: setHoveringSpenIconStyle1 pointerType: 10001 iconType:1 flag:0 pid:4141 uid:1000
12-22 14:41:50.178  4141  4331 D PointerIcon: setHoveringSpenCustomIcon IconType is same.1
12-22 14:41:50.228  4141 23872 E android.os.Debug: ro.product_ship = true
12-22 14:41:50.228  4141 23872 E android.os.Debug: ro.debug_level = 0x4f4c
12-22 14:41:50.228  4141 23872 E android.os.Debug: sys.mobilecare.preload = false
12-22 14:41:50.268  4141  4304 D ViewRootImpl: #1 mView = com.android.internal.policy.PhoneWindow$DecorView{40b8490 V.E...... R.....I. 0,0-0,0}
12-22 14:41:50.268  4141  4304 D ISSUE_DEBUG: InputChannelName : d74158e Application Error: com.hurryup.merchant
12-22 14:41:50.273  4141  4304 V MARsPolicyManager: handelAlertToastWindowStarted pkgName = android
12-22 14:41:50.273  4141  4304 D InputDispatcher: Focus entered window: 4141
12-22 14:41:50.273  4141  4331 D PointerIcon: setMouseIconStyle1 pointerType: 1001 iconType:101 flag:0 pid:4141 uid:1000
12-22 14:41:50.273  4141  4331 D PointerIcon: setMouseCustomIcon IconType is same.101

```

So found the culprit. OutOfMemoryError. What to do? 

My AndroidMenifest:
```
    <uses-permission android:name="android.permission.INTERNET"/>
    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION"/>
    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"/>
    <uses-permission android:name="android.permission.CAMERA"/>

```

**Comment by Sh1d0w** [comment]
Try to pass a quality param to compress the returned image eg, put 85. What is the real size of the image you are trying to pick and how much memory does your device has?

**Comment by Yunus0or1** [comment]
Samsung s5: Ram 2GB
Pic Size: 3MB
Actually the problem happened not for the first time. So that is why I picked image using multi_image_picker and then got filePath and using filePath I could access the image from Store. The latest version removed filePath so I could not do it. Any work around ?

**Comment by Sh1d0w** [comment]
What are you actually doing with the byte data in the loop? Can you provide more code?

**Comment by Yunus0or1** [comment]
```
  List<Asset> images = List<Asset>();
  List<ByteData> fileData = List<ByteData>(); 

 Future<void> displayAddDialog() async {

    try {

      setState(() {
        images = List<Asset>();
      });

      List<Asset> resultList;
      String error;

      try {
        resultList = await MultiImagePicker.pickImages(
            maxImages: 10, enableCamera: true);
      } on Exception catch (e) {
        error = e as String;
      }


      setState(() {
        images = resultList;
        if (error == null) error = 'No Error Dectected';
      });

      for (int i = 0; i < images.length; i++) {

        try {
          ByteData imageData = await images[i].getThumbByteData(600, 800,quality:20 );
          fileData.add(imageData);

        } catch (error) {
          print(error);
        }
      }


	  // Uploading images to server
      for (int i = 0; i < fileData.length; i++) {
        if (mounted) setState(() {});
        final imageUrl =
            await ClientFactory.getImageClient().resolveImage(fileData[i]);
        widget.onSelect(imageUrl);
      }

      fileData.clear();
    } catch (err) {
    } finally {
      loading = false;
      if (mounted) setState(() {});
    }
  }

```

**Comment by Yunus0or1** [comment]
Same issue I reported before:

https://github.com/Sh1d0w/multi_image_picker/issues/225

**Comment by Sh1d0w** [comment]
You are crashing the app with the code, because you are continuously requesting byte data and pushing it to array thus you are increasing the memory consumed by the app with each image size. This is wrong.

You should request the image data only when you need it like for uploading and then free the memory. If you need to display thumbs of the picked images you can use the AssetThumb as in the example app.

You need to be careful with image data and request only when you need it and free it immediately.

I am going to close this as there is not actual bug here.

**Comment by Yunus0or1** [comment]
So I actually found out the solution. If I even choose one image it would crash the app in lower spec Android phones. So I added two lines in AndroidManifest.xml 

```
<application
        android:hardwareAccelerated="false"
        android:largeHeap="true"
</application>
```

So whenever I choose images now, more ram will be focused on this task. And it does not crash. I think this should be included in documentation if no other solution works. I know largeHeap should not be used in every app.
